### PR TITLE
Add support for Jexl

### PR DESCRIPTION
**Pull request by czosel**
Jexl is the [Javascript Expression Language](https://github.com/TomFrost/Jexl).

### CODE REVIEW

**Comment by czosel** [comment]
@RunDevelopment Before I resolve the conflicts, is this something you‘d be interested in? Thanks in advance!

**Comment by czosel** [comment]
Conflicts are resolved.

File: tests/languages/jexl/operator_feature.test [hunk 1]
```diff
@@ -0,0 +1,33 @@
+-
++
+< <=
+> >=
+= ==
```
Comment (RunDevelopment) [inline tests/languages/jexl/operator_feature.test:5]: Is `=` even an operator? The doc doesn't mention it.

**Review by RunDevelopment — CHANGES_REQUESTED** [review]
Sorry for the delay @czosel.

I looked at Jexl and your language definition and there are few things that need to be addressed (nothing major). 

Also, please follow the quoting style we use. We use single quotes and token names are always quoted. See [CSS](https://github.com/PrismJS/prism/blob/master/components/prism-css.js) as an example.

---

I want to ask about the current status of the Jexl project. There hasn't been any activity for half a year now which is kinda worrying.

**Comment by czosel** [comment]
Hi @RunDevelopment, thanks for the detailed review! I'll address the points you mention over the coming days. 

Regarding the status of the project: It's a small language (just expressions), I'd see the fact that nothing happened over the last 6 months as a sign of stability :wink: I'm not involved with the Jexl project myself, but my company is a heavy user of Jexl and would potentially be committed to help maintaining it, should the need ever arise.

**Comment by RunDevelopment** [comment]
>  I'd see the fact that nothing happened over the last 6 months as a sign of stability

The lack of commits, yes. What worried me more was that nobody responds to new issues and PRs.

But it's good to know that there are people willing to step in.

File: components/prism-jexl.js [hunk 2]
```diff
@@ -0,0 +1,13 @@
+Prism.languages.jexl = {
+	string: /(["'])(?:\\(?:[\s\S])|(?!\1)[^\\])*\1/,
```
[the remark this answers is not in the document]
Comment (czosel) [reply components/prism-jexl.js:2]: Concerning `"Hello "user""`: This seems to be a mistake in the docs, it isn't [parsed by Jexl](https://czosel.github.io/jexl-playground/#/?context=%7B%7D&input=%22Hello%20%22user%22%22&transforms=%7B%7D). 

Concerning escaping: This seems to be a bug, `'\'` is evaluated as `\\'`, which is probably not intended (see [example](https://czosel.github.io/jexl-playground/#/?context=%7B%7D&input='%5C'&transforms=%7B%7D)).

File: components/prism-jexl.js [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+	transform: {
+		pattern: /\|\s*[a-zA-Zа-яА-Я_\u00C0-\u00D6\u00D8-\u00F6\u00F8-\u00FF$][a-zA-Zа-яА-Я0-9_\u00C0-\u00D6\u00D8-\u00F6\u00F8-\u00FF$]*/,
+		alias: "function",
+	},
+	function: /[a-zA-Zа-яА-Я_\u00C0-\u00D6\u00D8-\u00F6\u00F8-\u00FF$][a-zA-Zа-яА-Я0-9_\u00C0-\u00D6\u00D8-\u00F6\u00F8-\u00FF$]*\s*(?=\()/,
+	number: /\b(?:(?:\d*\.\d+)|\d+)/,
```
[the remark this answers is not in the document]
Comment (czosel) [reply components/prism-jexl.js:8]: It's indeed strange that `.4` is not a number in Jexl. I'm a bit confused about the suggested change though - wouldn't this treat `.4` as a number?

File: components/prism-jexl.js [hunk 4]
```diff
@@ -0,0 +1,13 @@
+Prism.languages.jexl = {
+	string: /(["'])(?:\\(?:[\s\S])|(?!\1)[^\\])*\1/,
+	transform: {
+		pattern: /\|\s*[a-zA-Zа-яА-Я_\u00C0-\u00D6\u00D8-\u00F6\u00F8-\u00FF$][a-zA-Zа-яА-Я0-9_\u00C0-\u00D6\u00D8-\u00F6\u00F8-\u00FF$]*/,
```
[the remark this answers is not in the document]
Comment (czosel) [reply components/prism-jexl.js]: Good idea - done.

File: tests/languages/jexl/operator_feature.test (hunk 1 above)
Comment (czosel) [reply tests/languages/jexl/operator_feature.test]: Good catch! Removed it.

File: components/prism-jexl.js [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+		alias: "function",
+	},
+	function: /[a-zA-Zа-яА-Я_\u00C0-\u00D6\u00D8-\u00F6\u00F8-\u00FF$][a-zA-Zа-яА-Я0-9_\u00C0-\u00D6\u00D8-\u00F6\u00F8-\u00FF$]*\s*(?=\()/,
+	number: /\b(?:(?:\d*\.\d+)|\d+)/,
+	operator: /[<>]=?|[!=]=?|-|\+|&&|\|\||\/\/?|[?:*^%]/,
```
[the remark this answers is not in the document]
Comment (czosel) [reply components/prism-jexl.js]: :+1: done.

File: components/prism-jexl.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (czosel) [reply components/prism-jexl.js]: I simplified the regex as suggested.

**Comment by czosel** [comment]
Hi @RunDevelopment 
Sorry this took a bit longer than expected. I was able to address most of your feedback - thanks for taking another look!

File: components/prism-jexl.js [hunk 6]
```diff
@@ -0,0 +1,14 @@
+Prism.languages.jexl = {
+	'string': /(["'])(?:\\[\s\S]|(?!\1)[^\\])*\1/,
+	'transform': {
+		pattern: /(\|)\s*[a-zA-Zа-яА-Я_\u00C0-\u00D6\u00D8-\u00F6\u00F8-\u00FF$][a-zA-Zа-яА-Я0-9_\u00C0-\u00D6\u00D8-\u00F6\u00F8-\u00FF$]*/,
```
Comment (RunDevelopment) [inline components/prism-jexl.js:4]: Token (typically) shouldn't contain leading or trailing spaces.

```suggestion
		pattern: /(\|\s*)[a-zA-Zа-яА-Я_\u00C0-\u00D6\u00D8-\u00F6\u00F8-\u00FF$][a-zA-Zа-яА-Я0-9_\u00C0-\u00D6\u00D8-\u00F6\u00F8-\u00FF$]*/,
```

File: components/prism-jexl.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (RunDevelopment) [reply components/prism-jexl.js]: It really isn't. I'm sorry, this is my bad. I assumed that `.4` (as in JS) is a number. In that case, both the original and my suggested pattern are incorrect. I'll suggest a new one.

[the remark this answers is not in the document]
Comment (RunDevelopment) [reply components/prism-jexl.js]: As to _why_ `.4` isn't a number: The Jexl parser [matches grammar elements before numbers](https://github.com/TomFrost/Jexl/blob/29d1ac0e1fef05ff1b6ad8d9449472ccb8cfb091/lib/Lexer.js#L199) and [`.` is a grammar element](https://github.com/TomFrost/Jexl/blob/29d1ac0e1fef05ff1b6ad8d9449472ccb8cfb091/lib/grammar.js#L15), so no number can start with `.`. Is that a bug? Probably...

Actually, should we highlight what Jexl is supposed to do or what it actually does in this case?

**Comment by RunDevelopment** [comment]
Please ignore the last review comment. The _"It really isn't. I'm sorry, this is my bad. I assumed that .4 (as in JS) is a number. In that case, both the original and my suggested pattern are incorrect. I'll suggest a new one."_ one. GitHub has a bug with its review comment system and I triggered that bug...

File: components/prism-jexl.js (hunk 6 above)
Comment (czosel) [reply components/prism-jexl.js]: Of course - fixed.

**Comment by czosel** [comment]
@RunDevelopment Should we still change something about the numbers?

File: components/prism-jexl.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (czosel) [reply components/prism-jexl.js]: I wouldn't consider it a bug, since Jexl doesn't claim that it handles numbers exactly the same way as JavaScript does.

**Comment by RunDevelopment** [comment]
I say we just leave it as is.

**Review by RunDevelopment — APPROVED** [review]

**Comment by RunDevelopment** [comment]
Thank you for contributing @czosel!

**Comment by czosel** [comment]
Awesome :tada: Thanks for the detailed review, and your work on Prism in general!

Do you already have a new release planned? :-)

**Comment by RunDevelopment** [comment]
Thanks.

I do want to make a new release soon-ish but not until a few of the more recent PRs get merged.

**Pull request merged**